redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt

```diff
@@ -18,20 +18,28 @@ package app.cash.redwood.treehouse
 import androidx.compose.runtime.MutableState
 import androidx.compose.runtime.mutableStateOf
 import kotlin.jvm.JvmInline
+import kotlinx.serialization.Contextual
 import kotlinx.serialization.Serializable
 import kotlinx.serialization.json.JsonArray
 import kotlinx.serialization.json.JsonElement
 import kotlinx.serialization.json.JsonPrimitive
 import kotlinx.serialization.json.booleanOrNull
 import kotlinx.serialization.json.doubleOrNull
+import kotlinx.serialization.json.intOrNull
 
 @Serializable
 public class StateSnapshot(
-  public val content: Map<String, List<JsonElement?>>,
+  public val content: Map<String, List<@Contextual Saveable>>,
 ) {
-  public fun toValuesMap(): Map<String, List<Any?>>? {
+  public fun toValuesMap(): Map<String, List<Any?>> {
     return content.mapValues { entry ->
-      entry.value.map { mutableStateOf(it.fromJsonElement()) }
+      entry.value.map {
+        if (it.isMutableState) {
+          mutableStateOf(it.value)
+        } else {
+          it.value.fromJsonElement()
+        }
```

Comment:
We could remove the need for the `Saveable` wrapper class if we wrote an adapter to convert between `MutableState` and `JsonElement`. Perhaps it isn't worth the effort though, considering we'll change up this mechanism in #1384.